spec/requests/api/v1/sessions/facebook_spec.rb

```diff
@@ -0,0 +1,100 @@
+require 'rails_helper'
+
+describe 'POST api/v1/users/facebook', type: :request do
+  let(:facebook_path) { api_v1_users_facebook_path }
+  let(:params) { { access_token: '12345678' } }
+  let(:user) { User.last }
+
+  before do
+    stub_request(:get, 'https://graph.facebook.com/me')
+      .with(query: hash_including(
+        'access_token': '12345678',
+        'fields': 'email,first_name,last_name,gender'
+      ))
+      .to_return(status: 200, body: '{
+          "email": "test@facebook.com",
+          "first_name": "Jhon",
+          "last_name": "Doe",
+          "gender": "male",
+          "id": "987654321"
+        }')
+  end
+
+  shared_examples 'valid client and access token' do
+    it 'returns a valid client and access token' do
+      post facebook_path, params: params, as: :json
+      token = response.headers['access-token']
+      client = response.headers['client']
+      expect(user.reload.valid_token?(token, client)).to be_truthy
+    end
+  end
+
+  context 'when user logs in for the first time' do
+    it 'returns a succesful response' do
+      post facebook_path, params: params, as: :json
+      expect(response).to have_http_status(:success)
+    end
+
+    it 'creates a user' do
+      expect do
+        post facebook_path, params: params, as: :json
+      end.to change(User, :count).by(1)
+    end
+
+    it 'creates user properly' do
+      post facebook_path, params: params, as: :json
+      expect(user.encrypted_password).to be_present
+      expect(user.name).to eq 'Jhon Doe'
+      expect(user.gender).to eq 'male'
+      expect(user.email).to eq 'test@facebook.com'
+    end
+
+    it_behaves_like 'valid client and access token'
+  end
+
+  context 'with already registered user with same email' do
+    before do
+      create(:user, email: 'test@facebook.com')
+    end
```

Comment:
Same as `let!(:user) { create :user, email: 'test@facebook.com' }`. Don't change this if you want, just to remind you a 1 line way.